libraries/Wire/src/utility/twi.c

```diff
@@ -846,9 +846,10 @@ i2c_status_e i2c_slave_write_IT(i2c_t *obj, uint8_t *data, uint16_t size)
   } else {
     // Check the communication status
     for (i = 0; i < size; i++) {
-      obj->i2cTxRxBuffer[i] = *(data + i);
+      obj->i2cTxRxBuffer[obj->i2cTxRxBufferSize + i] = *(data + i);
     }
 
+    size = size + obj->i2cTxRxBufferSize;
```

Comment:
Nitpick: Both lines could be simplified to 
`obj->i2cTxRxBufferSize += size;`

Reply:
Hi @ABOSTM ,
thank you for the quick response.
To describe my main Setup:
I have 1 Raspberry Pi <- I2C Master --- I2C Slave -> 1 STM Nucleo F746ZG(stm32duino) <- I2C Master --- I2C Slave -> 4 STM Nucleo F746ZG(STMCube with MotorControl)
So the STM running the stm32duino is working with 2 I2C Instances. One as a master, one as a slave.
For the first tests, I stopped all communication and working with the 4 other Nucleos and only had the RPI<-> stm32duino communication running. I also tested with an Arduino Mega 2560 as the Master to the stm32duino to be able to control if the stop condition is sent(i couldn't do it easily with the rpi).

How I tested:
I had an Oscilloscope attached to look at the lines, and for the issue with repeated start, I could see that the transmission was there and acknowledged, but it didn't arrive in my Arduino code. (I think the Issue here is that HAL_I2C_ListenCpltCallback() isn't called when there is no stop condition).
For the issue with the write, it looked like the Nucleo was sending only 1 byte of data back and then held SCL low.

So the master which sends the repeated start is the RPI or the Arduino Mega.

Sadly, I don't have any other boards to test the functionality on other boards of the STM family.

My codebase has been getting quite big, so I don't know if you would want to have a look at it, but I will try to get a smaller example of the issue.
What would be the best way of sharing it?

Reply:
**How to use it**
The useRestartCondition define in the Master code is for the issue with the restart condition,
For the other issue, in Slave there is respondMultiWrite.

To read data, I send via serial to the master as ASCII: `8,0,1,0\0A`
For the restart condition issue: if it works, you get on the Slave as Serial output:
received adress 0x1
transmitted data 0x5555

if it didn't work, the received adress line is missing.

For the other issue, if it didn't work, it will lock up the I2C bus. So a reboot of both devices is necessary afterward.

I didnt rewrite the Master code, so forgive me that its a little bit messy
**Master (Arduino Mega 2560)**
```#include <Arduino.h>
#include <Wire.h>

#define useRestartCondition true
#define WIRE_INTERFACE_HANDLE Wire
uint32_t noResponseCount = 0;
uint32_t writeFailedCount = 0;
uint32_t regAddrFailedCount = 0;
uint32_t errCount = 0;
uint32_t transactionCount = 0;
uint32_t lastPrint = 0;

uint8_t lastReadNum = 0;

void setup() {
  // initialize I2C
  WIRE_INTERFACE_HANDLE.begin();

  Serial.begin(115200);
  Serial.println("hello");
}

void sendReg(uint8_t address, uint8_t reg, uint16_t data) {
  uint8_t success;
  do {
    writeFailedCount++;
    // Start a Transmission to device with address 64
    WIRE_INTERFACE_HANDLE.beginTransmission(address);
    // Write reg&data to the device
    WIRE_INTERFACE_HANDLE.write(reg);
    WIRE_INTERFACE_HANDLE.write(data & 0xff);
    WIRE_INTERFACE_HANDLE.write((data >> 8) & 0xff);
    success = WIRE_INTERFACE_HANDLE.endTransmission();
  } while (success != 0);
  //subtract the one time which succeeded
  writeFailedCount--;
}
uint8_t readReg(uint8_t address, uint8_t reg, uint16_t *data) {
  uint8_t success;
  do {
    regAddrFailedCount++;
    // Start a Transmission to device with address 64
    WIRE_INTERFACE_HANDLE.beginTransmission(address);
    // Write reg&data to the device
    WIRE_INTERFACE_HANDLE.write(reg);
    success = WIRE_INTERFACE_HANDLE.endTransmission(!useRestartCondition);
  } while (success != 0);
  //subtract the one time which succeeded
  regAddrFailedCount--;
  
  lastReadNum = WIRE_INTERFACE_HANDLE.requestFrom(address, (uint8_t) 2);    // request 2 bytes from slave device #64

  if (lastReadNum != 2) {
    Serial.print("DATALEN:");
    Serial.println(lastReadNum);

    while(WIRE_INTERFACE_HANDLE.available()) {
      int c = Wire.read(); // receive a byte as character
      Serial.print(c);         // print the character
      Serial.print(',');         // print the character
    }

    return 1;
  }

  //receiving
  if (WIRE_INTERFACE_HANDLE.available() >= 2) {
    uint16_t r1 = WIRE_INTERFACE_HANDLE.read();
    uint16_t r2 = WIRE_INTERFACE_HANDLE.read();
    *data = ((r2 << 8) | r1);
    return 0;
  }
  return 2;
}

void loop() {
  while (Serial.available() > 0) {
    int addr = Serial.parseInt();
    int rw = Serial.parseInt();
    int reg = Serial.parseInt();
    int data = Serial.parseInt();
    if (Serial.read() == '\n') {
      if ((rw == 0 || rw == 1) && reg >= 0 && reg <= 255 && data >= 0 && (uint32_t)data <= 0xffffffff) { // (addr == 0x8) && 
        if (rw) {
          // write
          Serial.print("Sending:");
          sendReg(addr, reg, (uint16_t) data);
          Serial.print(addr, HEX);
          Serial.print("/");
          Serial.print(reg, HEX);
          Serial.print("/");
          Serial.println(data, HEX);
        } else {
          Serial.print("Receiving:");
          uint8_t err = readReg(addr, reg, (uint16_t*) &data);
          Serial.print(addr, HEX);
          Serial.print("/");
          Serial.print(reg, HEX);
          Serial.print("/");
          Serial.print(data, HEX);
          Serial.print("/");
          Serial.print(data);
          Serial.print("/");
          Serial.println(err, HEX);
        }
      } else {
        Serial.print("wrong param");
        Serial.print(addr, HEX);
         Serial.print("/");
        Serial.print(rw, HEX);
        Serial.print("/");
        Serial.print(reg, HEX);
        Serial.print("/");
        Serial.println(data, HEX);
      }
    }
  }
}```
**Slave (stm32duino)**
```#include <Arduino.h>
#include <Wire.h>

#define respondMultiWrite false

// Use Wire2, because Wire is used for other communication
//            SDA  SCL
TwoWire Wire2(PB11, PB10);

void i2cReceiveEvent(int numBytes);
void i2cRequestEvent();

volatile uint8_t i2cCommand; // register adress

void setup() {
  Serial.begin(115200);
  
  Wire2.begin(0x08, true); // boolean is general call enable/disable
  Wire2.onReceive(i2cReceiveEvent);  // register event
  Wire2.onRequest(i2cRequestEvent);  // register event
  Serial.println("Started arduino firmware!");
}

void loop() {
  // put your main code here, to run repeatedly:
}

// function that executes whenever data is received from master
// this function is registered as an event, see setup()
void i2cReceiveEvent(int numBytes) { 
  // there are two cases:
  // 1. user sends a 16-bit write command, then the transmission is 1+2 bytes long
  // 2. user sends a 16-bit read command, then the transmission is 1 byte long. After that a 2 byte request is received.

  // case 1
  if ((3 == numBytes) && (Wire2.available() > 0)) {
    i2cCommand = Wire2.read();  // First byte
    uint8_t b = Wire2.read();   // Second byte
    uint8_t c = Wire2.read();   // Third byte

    // Get the I2C Value
    uint16_t i2cVal = ((c << 8) & 0xFF00) | (b & 0xFF);

    Serial.print("recevied adress 0x");
    Serial.print(i2cCommand, HEX);
    Serial.print(" with data 0x");
    Serial.println(i2cVal, HEX);
  }

  // case 2
  // User issues i2c 16-bit register read command
  if ((1 == numBytes) && (Wire2.available() > 0)) {
    i2cCommand = Wire2.read();
    Serial.print("received adress 0x");
    Serial.println(i2cCommand, HEX);
  }
}

// function that executes whenever data is requested by master
// this function is registered as an event, see setup()
void i2cRequestEvent() {

  if (respondMultiWrite) {
    // The way i would like it to also work
    int16_t regVal = 0x5555;
    //Wire2.write(regVal & 0xFF);         // Send data back to the Master
    //Wire2.write((regVal >> 8) & 0xFF);  // Send data back to the Master
  } else {
    // The old way to write data
    const uint8_t data[2] = {0x55, 0x55}; 
    const uint8_t* dataPtr = (const uint8_t*) &data;
    Wire2.write(dataPtr, (size_t) 2);
  }  

  Serial.print("transmitted data 0x");
  Serial.println(regVal, HEX);
}```